polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -609,16 +609,18 @@ bool ArgumentPromotionPass::isCandidateCallable(
     return false;
   }
 
+  // Check candidate is a SYCL KernelObjFunction.
+  if (!funcKernelInfo.isKernelFuncObjFunction(funcOp)) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "not a candidate: not a SYCL KernelObjFunction\n");
+    return false;
+  }
+
   // Ensure all the call sites for this function are either in a GPU kernel or
   // in a function that is called directly by a GPU kernel.
   // TODO: Could generalize by checking that the call chain from the GPU kernel
   // are all candidates.
-  Optional<unsigned> maxDepth =
-      funcKernelInfo.getMaxDepthFromAnyGPUKernel(funcOp);
-  assert(maxDepth.has_value() &&
-         "Expecting func to be called from a GPU kernel");
-  assert(maxDepth.value() != 0 && "Expecting func is not itself a GPU kernel");
-  if (maxDepth.value() > 2) {
+  if (!funcKernelInfo.isCalledDirectlyFromKernel(callableOp)) {
```

Comment:
`GPU kernel -> B -> callableOp`
This is allowed before, as `maxDepth.value()` would be 2, the call site (call to callableOp) is in a function (B) that is called directly by a GPU kernel.

With the current implementation of SemaSYCL, a SYCL KernelObjFunction should always met this condition we are checking here. We would only adding this check to avoid the non SYCL KernelObjFunction.

Reply:
Nice. We can drop the function in that case altogether